src/zycast.c

```diff
@@ -0,0 +1,336 @@
+{
+	fprintf(stderr, "ERR: %s: %s\n", msg, errno ? strerror(errno) : "unknown");
+	exit(EXIT_FAILURE);
+}
+
+static void *map_input(const char *name, size_t *len)
+{
+	struct stat stat;
+	void *mapped;
+	int fd;
+
+	fd = open(name, O_RDONLY);
+	if (fd < 0)
+		return NULL;
+	if (fstat(fd, &stat) < 0) {
+		close(fd);
+		return NULL;
+	}
+	*len = stat.st_size;
+	mapped = mmap(NULL, stat.st_size, PROT_READ, MAP_SHARED, fd, 0);
+	if (close(fd) < 0) {
+		(void) munmap(mapped, stat.st_size);
+		return NULL;
+	}
+	return mapped;
+}
+
+static uint16_t chksum(uint8_t *p, size_t len)
+{
+	int i;
+	uint32_t sum = 0;
+
+	for (i = 0; i < len; i++)
+		sum += *p++;
+	return (uint16_t)((uint32_t)(sum >> 16) + sum);
+}
+
+static int pushimage(void *file, struct zycast_t *phdr)
+{
+	uint32_t count = 0;
+	uint32_t len = ntohl(phdr->flen);
+	uint32_t plen = CHUNK;
+
+	while (!exiting && len > 0) {
+		if (len < CHUNK)
+			plen = len;
+		phdr->plen = htonl(plen);
+		phdr->pid = htonl(count++);
+		phdr->chksum = htons(chksum(file, plen));
+		if (send(sockfd, phdr, HDRSIZE, MSG_MORE | MSG_DONTROUTE) < 0)
+			errexit("send(phdr)");
+		if (send(sockfd, file, plen, MSG_DONTROUTE) < 0)
+			errexit("send(payload)");
+		file += plen;
+		len -= plen;
+
+		/* No need to kill the network. The target can't
+		 * process packets as fast as we send them anyway.
+		 */
+		usleep(pktdelay);
+	}
+	return 0;
+}
+
+static void sig_handler(int signo)
+{
+	if (signo == SIGINT)
+		exiting = 1;
+}
+
+static void usage(const char *name)
+{
+	fprintf(stderr, "Usage:\n");
+	fprintf(stderr, " %s [options]\n", name);
+	fprintf(stderr, "Options:\n");
+	fprintf(stderr, "\t-i interface            outgoing interface for multicast packets\n");
+	fprintf(stderr, "\t-t delay                interpacket delay in milliseconds\n");
+	fprintf(stderr, "\t-f rasimage             primary firmware image\n");
+	fprintf(stderr, "\t-b backupimage          secondary fimrware image (if supported)\n");
+	fprintf(stderr, "\t-d rom                  data for the \"rom\" or \"data\" partition\n");
+	fprintf(stderr, "\t-r romd                 data for the \"rom-d\" partition\n");
+#ifdef DO_BOOTBASE
+	fprintf(stderr, "\t-u bootloader           flash new bootloader\n");
+	fprintf(stderr, "\nWARNING: bootloader upgrades are dangerous.  DON'T DO IT!\n");
+#endif
+	fprintf(stderr, "\nNOTE: some bootloaders will flash a rasimage to both primary and\n");
+	fprintf(stderr, "secondary firmare partitions\n");
+	fprintf(stderr, "\nExample:\n");
+	fprintf(stderr, " %s -i eth1 -t 20 -f openwrt-initramfs.bin\n\n", name);
+	if (sockfd >= 0)
+		close(sockfd);
+	exit(EXIT_FAILURE);
+}
+
+#define ADD_IMAGE(nr) \
+	hdr.images |= BIT(nr); \
+	file[nr] = map_input(optarg, &len[nr]); \
+	if (!file[nr]) \
+		errexit(optarg)
+
+int main(int argc, char **argv)
+{
+	void *file[_MAX_IMAGETYPE] = {};
+	size_t len[_MAX_IMAGETYPE] = {};
+	struct zycast_t hdr = {
+		.magic = htonl(MAGIC),
+		.cc    = {'F', 'F' },
+		.flags = FLAG_SET_DEBUG,
```

Comment:
Since you're using at least this C99 feature, you can also declare loop variables inside 'for' (`for (int i = 0; ...)`).
Or even use C11, if you like. You can define the standard you want to use in CMakeLists.txt using the second argument in `FW_UTIL()`

Reply:
Thanks a lot for the review!  I've fixed up all the other issues, and a lot more speling issues.  Including the misspelled port number in the protocol description....

But I'll leave the lool var declarations as-is unless you insist.   I pick features I like and ignore the rest, hoping that no one will notice that I never read any C standards at all :-)

Reply:
Maybe I'm just too young to understand the appeal of C89 ;-)